lessons/advanced/typespec.md

```diff
@@ -0,0 +1,127 @@
+
+However in some cases specification is going to be quite big and complicated. If you would like to reduce complexity, you want to introduce custom type definition. Elixir has `@type` annotation for that. In the other hand, Elixir is still dynamic language. That means all information about type will be ignored by compiler, but could be used by other tools.   
+
+## Specification
+
+If you have experience with Java or Ruby you could think about specification as an `interface`. Specification defines what should be type of function parameters and return value.
+
+To define input and output types we use `@spec` directive placed right before function definition and taking as a `params` name of function, list of parameter types, and after `::` type of return value.  
+
+Let's take a look at example:
+
+```elixir
+@spec sum_product(integer) :: integer
+def sum_product(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+end
+```
+
+Everything looks ok and when we call valid result will be return, but function `Enum.sum` returns `number` not `integer` as we expected in `@spec`. It could be source of bugs! There are tools like Dialyzer to static analysis of code that helps us to find this type of bugs. We will talk about them in another lesson.
+ 
+## Custom types
+
+Writing specification is nice, but sometimes our functions works with more complex data structures than simple numbers or collections. In that cases definition in `@spec` could be hard to understand and very hard to change by other developers. Sometimes functions needs to take many parameters or return complex data. Long parameters list is one of bad smells in code. In object oriented languages like Ruby or Java we could easily define classes that helps us to solve this problems. Elixir hasn't classes but because is easy to extends that we could define our types.
+  
+Out of box Elixir contains some basic types like `integer` or `pid`. You  can find full list of available types in [documentation](http://elixir-lang.org/docs/stable/elixir/typespecs.html#types-and-their-syntax).
+ 
+### Defining custom type
+  
+Let's modify our `sum_times` function and introduce some extra params:
+
+```elixir
+@spec sum_times(integer, %Examples{first: integer, last: integer}) :: integer
+def sum_times(a, params) do
+    for i <- params.first..params.last do
+        i
+    end
+       |> Enum.map(fn el -> el * a end)
+       |> Enum.sum
+       |> round
+end
+```
+
+We introduce struct in `Examples` module that contains two fields `first` and `last`. That is simpler version of struct from `Range` module. About structures we talk when discussing [modules](lessons/basics/modules/#structs). Lets imagine that we need to specification with `Examples` struct many times. It will be very annoying and could be source of bugs. Solution of this problem is `@type`.
```

Comment:
`We introduce struct in `Examples` module that contains two fields `first` and `last`.`

-->

`We introduced a struct in our `Examples` module that contained two fields: ``` `first`, and `last`. ```

===================

`About structures we talk when discussing [modules](lessons/basics/modules/#structs).`

-->

```We will talk about `structs` when we get into discussing [modules](lessons/basics/modules/#structs).```

====================

I cant reason through what this is saying:

```Lets imagine that we need to specification with `Examples` struct many times. It will be very annoying and could be source of bugs. Solution of this problem is `@type`.```

===========================

```It will be very annoying and could be source of bugs. Solution of this problem is `@type`.```

-->

```It would be very annoying and could be a source of bugs. A solution to this problem is `@type`.```